src/main/java/io/gravitee/policy/json2xml/JsonToXmlTransformationPolicy.java

```diff
@@ -15,74 +15,103 @@
+public class JsonToXmlTransformationPolicy extends JsonToXmlTransformationPolicyV3 implements Policy {
 
-    private static final String UTF8_CHARSET_NAME = "UTF-8";
-    static final String CONTENT_TYPE = MediaType.APPLICATION_XML + ";charset=" + UTF8_CHARSET_NAME;
+    public JsonToXmlTransformationPolicy(final JsonToXmlTransformationPolicyConfiguration configuration) {
+        super(configuration);
+    }
 
-    /**
-     * Json to xml transformation configuration
-     */
-    private final JsonToXmlTransformationPolicyConfiguration configuration;
+    @Override
+    public String id() {
+        return "json-xml";
+    }
 
-    public JsonToXmlTransformationPolicy(final JsonToXmlTransformationPolicyConfiguration configuration) {
-        this.configuration = configuration;
+    @Override
+    public Completable onRequest(final HttpExecutionContext ctx) {
+        return ctx
+            .request()
+            .onBody(
+                bodyUpstream ->
+                    bodyUpstream
+                        .flatMap(buffer -> transformToXml(buffer, CharsetHelper.extractCharset(ctx.request().headers())))
+                        .doOnSuccess(xmlBuffer -> applyHeaders(ctx.request().headers(), xmlBuffer))
+            )
+            .onErrorResumeNext(throwable -> interruptWithBadRequest(ctx, throwable));
+    }
+
+    @Override
+    public Completable onResponse(final HttpExecutionContext ctx) {
+        return ctx
+            .response()
+            .onBody(
+                bodyUpstream ->
+                    bodyUpstream
+                        .flatMap(buffer -> transformToXml(buffer, CharsetHelper.extractCharset(ctx.response().headers())))
+                        .doOnSuccess(xmlBuffer -> applyHeaders(ctx.response().headers(), xmlBuffer))
+            )
+            .onErrorResumeNext(throwable -> interruptWithBadRequest(ctx, throwable));
     }
 
-    @OnResponseContent
-    public ReadWriteStream onResponseContent(Response response, PolicyChain chain) {
-        if (configuration.getScope() == null || configuration.getScope() == PolicyScope.RESPONSE) {
-            Charset charset = CharsetHelper.extractCharset(response.headers());
+    @Override
+    public Completable onMessageRequest(MessageExecutionContext ctx) {
+        return ctx
+            .request()
+            .onMessage(
+                message -> transformToXml(message.content(), CharsetHelper.extractCharset(ctx.request().headers())).map(message::content)
+            );
+    }
 
-            return TransformableResponseStreamBuilder.on(response).chain(chain).contentType(CONTENT_TYPE).transform(map(charset)).build();
-        }
-        return null;
+    @Override
+    public Completable onMessageResponse(MessageExecutionContext ctx) {
+        return ctx
+            .response()
+            .onMessage(
+                message -> transformToXml(message.content(), CharsetHelper.extractCharset(ctx.response().headers())).map(message::content)
```

Comment:
Should we update header to add some sort of content type ?

Reply:
Done